crates/red_knot_python_semantic/resources/mdtest/scopes/global.md

```diff
@@ -0,0 +1,200 @@
+# `global` references
+
+## Implicit global in function
+
+A name reference to a never-defined symbol in a function is implicitly a global lookup.
+
+```py
+x = 1
+
+def f():
+    reveal_type(x)  # revealed: Unknown | Literal[1]
+```
+
+## Explicit global in function
+
+```py
+x = 1
+
+def f():
+    global x
+    reveal_type(x)  # revealed: Unknown | Literal[1]
+```
+
+## Unassignable type in function
+
+```py
+x: int = 1
+
+def f():
+    y: int = 1
+    # error: [invalid-assignment] "Object of type `Literal[""]` is not assignable to `int`"
+    y = ""
+
+    global x
+    # @Todo(error: [invalid-assignment] "Object of type `Literal[""]` is not assignable to `int`")
+    x = ""
+```
+
+## Nested intervening scope
+
+TODO this should give the outer module type of `Unknown | Literal[1]`, not `Literal[""]`
+
+```py
+x = 1
+
+def outer():
+    x = ""
+
+    def inner():
+        global x
+        reveal_type(x)  # revealed: Unknown | Literal[""]
+```
+
+## Narrowing
+
+An assignment following a `global` statement should narrow the type in the local scope after the
+assignment. The revealed type of `Literal[1]` here is in line with [pyright], while [mypy] reports
+`builtins.int`.
+
+```py
+x: int | None
+
+def f():
+    global x
+    x = 1
+    reveal_type(x)  # revealed: Literal[1]
+```
+
+This related case is adapted from a `mypy` [test][t] with the comment:
+
+> This is unsafe, but we don't generate an error, for convenience. Besides, this is probably a very
+> rare case.
+
+```py
+g: str | None
+
+def f():
+    global g
+    g = "x"
+    def nested() -> str:
+        return g
+```
```

Comment:
This is a useful test case, but despite happening to use a `global` statement, it's not related to `global`. The same case could be constructed without `global`, just making `g` a local in `f` and using an `assert` (or `if` with early return, or whatever) to narrow its type instead. In other words, the test is really about how nested functions can or cannot make use of narrowed types in the outer scope. Which is a topic we do need to address, but it's not related to `global`. So I don't think we should include this test here.
```suggestion
```